markets/retrievaladapter/provider.go

```diff
@@ -38,6 +38,53 @@ func NewRetrievalProviderNode(maddr dtypes.MinerAddress, secb sectorblocks.Secto
 	return &retrievalProviderNode{address.Address(maddr), secb, pp, full}
 }
 
+func (rpn *retrievalProviderNode) GetRetrievalPricingInput(ctx context.Context, pieceCID cid.Cid, storageDeals []abi.DealID) (retrievalmarket.PricingInput, error) {
+	resp := retrievalmarket.PricingInput{}
+
+	head, err := rpn.full.ChainHead(ctx)
+	if err != nil {
+		return resp, xerrors.Errorf("failed to get chain head: %s", err)
+	}
+
+	for _, dealID := range storageDeals {
+		ds, err := rpn.full.StateMarketStorageDeal(ctx, dealID, head.Key())
+		if err != nil {
+			return resp, xerrors.Errorf("failed to look up deals on chain: %w", err)
+		}
+		if ds.Proposal.VerifiedDeal {
+			resp.VerifiedDeal = true
+		}
+
+		if ds.Proposal.PieceCID.Equals(pieceCID) {
+			resp.PieceSize = ds.Proposal.PieceSize.Unpadded()
+		}
+	}
```

Comment:
I'm not fully grasping this logic. This takes a list of deals, and scans through them, never breaking out of the loop. I think this leads to inconsistent data, for example, if we have 3 deals [0..2]:

1. If 1 was verified, but 2 isn't, we will run with verified=true.
2. If 1 didn't match the piece CID, but 2 does, we will take the piece size of 2.

This seems a bit off.

Reply:
@raulk 

1. These are storage deals made for the `payloadCID` the retrieval client is looking for.
2. For the `VerifiedDeal ` flag in the `PricingInput`, I am looking to answer the question "does there exist any verified storage deal for this `payloadCID` " ?
3. Now, as per the current data model in markets, a `payloadCID` can belong to many Pieces and each Piece can correspond to multiple storage deals for that Piece, each deal being sealed in a different sector. We first select a Piece we want to retrieve the payload from (prioritising a Piece which has a storage deal on an Unsealed sector) and then pass the PieceCID to this function. Now, since the storage deals could be for different pieces(with these different pieces containing the same `payloadCID`), I wanna match the PieceCIDs to then determine the PieceSize for the retrieval deal. 

Not breaking the loop here after we've determined both pieces of information is a mistake, that can be fixed. Does this make sense ?

Reply:
cc @dirkmc for the Markets context.